### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by kovetskiy**

### CODE REVIEW

File: tests.sh [hunk 1]
```diff
@@ -1,5 +1,21 @@
 #!/bin/bash
 
+
+# Current test session
+TEST_ID=""
+
+# Verbosity level
+TEST_VERBOSE=0
+
+# Assertations counter
```
Comment (seletskiy) [inline tests.sh]: Assertions

File: tests.sh [hunk 2]
```diff
@@ -32,65 +48,90 @@ tests_assert_equals() {
     TEST_ASSERTS=$(($TEST_ASSERTS+1))
 }
 
-# Function tests_assert_stdout evaluates given command and compare it's stdout
+# Function tests_assert_stdout compare last evaluated command stdout
```
Comment (seletskiy) [inline tests.sh]: compares

File: tests.sh [hunk 3]
```diff
@@ -32,65 +48,90 @@ tests_assert_equals() {
...
-    tests_assert_stdout_re "$(tests_quote_re "$expected")" "${@}"
+    tests_assert_stdout_re "$(tests_quote_re "$expected")"
 }
 
-# Function tests_assert_re evaluates command and checks that it's output
+# Function tests_assert_re checks that last evaluated command output
 # (stdout or stderr) matches regexp.
 # Args:
-#   $1: stdout|stderr
+#   $1: stdout|stderr|file
```
Comment (seletskiy) [inline tests.sh]: file ->  &lt;filename>

File: tests.sh [hunk 4]
```diff
@@ -32,65 +48,90 @@ tests_assert_equals() {
...
+#   $2: stdout|stderr|string|file (actual)
+tests_diff() {
+    local expected_target="$1"
+    local actual_target="$2"
+    shift 2
+
+    if [ -f $expected_target ]; then
+        expected_file=$expected_target
+    else
+        expected_file=<(echo "$expected_target")
```
Comment (seletskiy) [inline tests.sh:106]: `expected_file="$expected_target"`?

File: tests.sh [hunk 5]
```diff
@@ -176,27 +202,42 @@ tests_debug() {
...
-# }}}
 
-# Internal Code {{{
+# Function tests_do evaluates specified string
+# Args:
+#   ${@}: string to evaluate
 tests_do() {
     tests_debug "$ ${@}"
 
+    echo >$TEST_STDOUT
```
Comment (seletskiy) [inline tests.sh]: Just `>$TEST_STDOUT`.

File: tests.sh [hunk 6]
```diff
@@ -176,27 +202,42 @@ tests_debug() {
...
+    echo >$TEST_EXITCODE
+
     {
         if [ $TEST_VERBOSE -lt 2 ]; then
-            tests_eval "${@}" > /dev/null 2>/dev/null
+            tests_eval "${@}" > $TEST_STDOUT 2> $TEST_STDERR
         fi
 
         if [ $TEST_VERBOSE -lt 3 ]; then
             tests_eval "${@}" > /dev/null
```
Comment (seletskiy) [inline tests.sh]: Forgot to remove `tests_eval "${@}" > /dev/null`?

File: tests.sh (hunk 4 above)
Comment (kovetskiy) [inline tests.sh:106]: and expected_file will be string, not fd.

File: tests.sh [hunk 7]
```diff
@@ -176,27 +202,42 @@ tests_debug() {
...
+            tests_eval "${@}" \
+                2> >(tee $TEST_STDERR 1>&2) \
+                1> >(tee $TEST_STDOUT > /dev/null)
         fi
 
         if [ $TEST_VERBOSE -ge 4 ]; then
-            tests_eval "${@}"
+            tests_eval "${@}" \
+                2> >(tee $TEST_STDERR 1>&2) \
```
Comment (seletskiy) [inline tests.sh]: BTW, it's all redirected to stdout later (235 line), so it's useless to redirect it to stderr there.

File: tests.sh [hunk 8]
```diff
@@ -301,12 +342,22 @@ tests_get_last() {
...
 tests_verbose() {
-    export TEST_VERBOSE=$1
+    TEST_VERBOSE=$1
 }
 
 tests_init() {
-    export TEST_ID="$(mktemp -t -d tests.XXXX)"
+    TEST_ID="$(mktemp -t -d tests.XXXX)"
+
+    TEST_STDERR="$TEST_ID/stderr"
```
Comment (seletskiy) [inline tests.sh:363]: What do you think about keeping stderr/stdout/exitcode separately for each executed command? Should be easier to debug, I guess.

File: tests.sh [hunk 9]
```diff
@@ -264,7 +305,7 @@ tests_run_all() {
                 echo
                 cat $stdout
                 rm -f $stdout
-                echo "$file" > .last-testcase
+                tests_set_last "$file"
```
Comment (seletskiy) [inline tests.sh:308]: Add deletion of .last-testcase file if all tests are passed.

**Pull request merged**